File: server/app/py_aco/graph.py
# import json
import math
import random


class Graph:
    """
    ノードやエッジ，フェロモンなどを管理する

    Attributes
    ----------
    params: Parameters
        ACOを設定するすべてのパラメータを保存する
    coordinates: list of tuple of float
        ノードの座標を格納する
    length_edge: list of int
        エッジの長さを格納する
    pheromone_edge: list of int
        現在蓄積されているエッジのフェロモン量を格納する
    next_pheromone_edge: list of int
        最新のサイクルで分泌されたフェロモン量を格納する
    heuristics_edge: list of int
        エッジのヒューリスティック値を格納する
    """

    def __init__(self, parameters, path):
        """
        Parameters
        ----------
        parameters: Parameters
            ACOを設定するすべてのパラメータを保存する
        """
        self.params = parameters
        num_node = self.params.num_node
        self.coordinates = [None] * num_node
        self.length_edge = [[0] * num_node for _ in range(num_node)]
        self.pheromone_edge = [[0] * num_node for _ in range(num_node)]
        self.next_pheromone_edge = [[0] * num_node for _ in range(num_node)]
        self.heuristics_edge = [[0] * num_node for _ in range(num_node)]
        self.__prepare_graph(path)

    def __prepare_graph(self, path):
        """
        受け取ったグラフ構造の情報を読み取る
        path: str
            グラフ構造を記述したファイルのパス
        """
        # with open(path, 'r') as f:
        #     file = json.load(f)
        # layouts = file.get("layouts", {})
        # nodes = layouts.get("nodes", {})
        # for node in nodes:
        #     node_id = node.get("id", 0)
        #     node_x = node.get("x", 0.0)
        #     node_y = node.get("y", 0.0)
        #     self.coordinates[node_id] = (node_x, node_y)
        # edges = layouts.get("edges", {})
        # for edge in edges:
        #     source = edge.get("source", 0)
        #     target = edge.get("target", 0)
        #     length = edge.get("length", 0)
        #     self.length_edge[source][target] = length
        #     self.length_edge[target][source] = length
        for i in range(self.params.num_node):
            self.coordinates[i] = (random.uniform(
                0, 1000), random.uniform(0, 1000))
        for i in range(self.params.num_node):
            for j in range(self.params.num_node):
                self.length_edge[i][j] = Graph.__calc_edge_length(
                    self.coordinates[i], self.coordinates[j])
        for i in range(self.params.num_node):
            for j in range(self.params.num_node):
                if i < j:
                    self.pheromone_edge[i][j] = self.params.q / \
                        self.length_edge[i][j]
                    self.pheromone_edge[j][i] = self.params.q / \
                        self.length_edge[i][j]
        for i in range(self.params.num_node):
            for j in range(self.params.num_node):
                if i < j:
                    h = self.params.q / self.length_edge[i][j]
                    self.heuristics_edge[i][j] = h
                    self.heuristics_edge[j][i] = h

    def reset_graph(self):
        self.next_pheromone_edge = [
            [0] * self.params.num_node for _ in range(self.params.num_node)]

    def reset_pheromones(self):
        """
        エッジに蓄積したフェロモンをすべてリセットする
        """
        for i in range(self.params.num_node):
            for j in range(self.params.num_node):
                if i < j:
                    self.pheromone_edge[i][j] = \
                        self.params.q / self.length_edge[i][j]
                    self.pheromone_edge[j][i] = \
                        self.params.q / self.length_edge[j][i]

    @staticmethod
    def __calc_edge_length(p1, p2):
        dx = (p1[0] - p2[0]) ** 2
        dy = (p1[1] - p2[1]) ** 2
        return math.sqrt(dx + dy)

File: server/app/aco/graph.go
package aco

import (
	"math"
)

type Graph struct {
	Params             Parameter
	Coordinates        []Point
	Distances          [][]float64
	TrailPheromones    [][]float64
	NewTrailPheromones [][]float64
	HeuristicValues    [][]float64
}

type Point struct {
	X float64
	Y float64
}

func NewGraph(params Parameter, points []Point) *Graph {

	graph := &Graph{}
	graph.Params = params
	graph.Coordinates = make([]Point, graph.Params.NumNodes)
	graph.Distances = make([][]float64, graph.Params.NumNodes)
	graph.TrailPheromones = make([][]float64, graph.Params.NumNodes)
	graph.NewTrailPheromones = make([][]float64, graph.Params.NumNodes)
	graph.HeuristicValues = make([][]float64, graph.Params.NumNodes)

	for i := 0; i < graph.Params.NumNodes; i++ {
		graph.Distances[i] = make([]float64, graph.Params.NumNodes)
		graph.TrailPheromones[i] = make([]float64, graph.Params.NumNodes)
		graph.NewTrailPheromones[i] = make([]float64, graph.Params.NumNodes)
		graph.HeuristicValues[i] = make([]float64, graph.Params.NumNodes)
	}

	graph.prepareGraph(points)

	return graph
}

func (g *Graph) prepareGraph(points []Point) {
	g.Coordinates = points
	for i := 0; i < g.Params.NumNodes; i++ {
		for j := 0; j < g.Params.NumNodes; j++ {
			g.Distances[i][j] = calcDistance(
				g.Coordinates[i], g.Coordinates[j],
			)
		}
	}

	for i := 0; i < g.Params.NumNodes; i++ {
		for j := 0; j < g.Params.NumNodes; j++ {
			if i < j {
				distance := g.Distances[i][j]
				g.TrailPheromones[i][j] = float64(g.Params.Q) / distance
				g.TrailPheromones[j][i] = float64(g.Params.Q) / distance
				g.HeuristicValues[i][j] = float64(g.Params.Q) / distance
				g.HeuristicValues[j][i] = float64(g.Params.Q) / distance
			}
		}
	}
}

func (g *Graph) resetGraph() {
	g.NewTrailPheromones = make([][]float64, g.Params.NumNodes)
	for i := range g.NewTrailPheromones {
		g.NewTrailPheromones[i] = make([]float64, g.Params.NumNodes)
	}
}

func (g *Graph) resetPheromones() {
	for i := 0; i < g.Params.NumNodes; i++ {
		for j := 0; j < g.Params.NumNodes; j++ {
			if i < j {
				distance := g.Distances[i][j]
				g.TrailPheromones[i][j] = float64(g.Params.Q) / distance
				g.TrailPheromones[j][i] = float64(g.Params.Q) / distance
			}
		}
	}
}

func calcDistance(p1, p2 Point) float64 {
	dx := p1.X - p2.X
	dy := p1.Y - p2.Y
	return math.Sqrt(dx*dx + dy*dy)
}
